### ISSUE TITLE
facet_wrap() and facet_grid() with a date variable is broken

### ISSUE THREAD

**Issue by paleolimbot** [tidyverse/ggplot2#3313]
The revdeps (#3303) identified that using `facet_grid()` or `facet_wrap()` with a date variable is broken. This probably affects other S3 vectors as well.

``` r
library(ggplot2)
df <- data.frame(date = as.Date("2019-01-01"), x = 1, y = 1)
p <- ggplot(df, aes(x, y)) + geom_point()
p + facet_grid(vars(date))
#> Error in scale_apply(layer_data, x_vars, "train", SCALE_X, x_scales):
p + facet_wrap(vars(date))
#> Error in scale_apply(layer_data, x_vars, "train", SCALE_X, x_scales):
```

The error is in `Facet(Grid|Wrap)$map_data()`, where there is a join between the layout and the panels. In one of these, `date` is a factor, and in the other, it is an S3 `Date`. This changed because of different behaviour between `join_keys()` and its previous **plyr** counterpart.

**Comment by clauswilke** [comment]
@yutannihilation Do you have time to look into this issue?

**Comment by yutannihilation** [comment]
Sure, will do.

**Comment by yutannihilation** [comment]
Quickly confirmed #3327 fixes this issue. Sorry, I'm afraid I won't be very responsive until the end of this month...

**Comment by paleolimbot** [comment]
I would be happy to PR a test for this if that is helpful.

**Comment by thomasp85** [comment]
I don’t think it makes sense to test for some arbitrary ordering as long as we don’t break something in revdeps... we will probably have to actively break this once we move to rbinding through vctrs at is is unlikely to follow the arbitrary rules of plyr, but it is best to break only once

**Comment by yutannihilation** [comment]
Though ordering is just an implementational detail, I think it makes sense to ensure facetting by Date is not broken.

**Comment by thomasp85** [comment]
I’m simply weary because the old ordering (which is now replicated) is logically wrong. Setting up a unit test may give it some undeserved credibility

**Comment by yutannihilation** [comment]
Yeah, I agree that we don't need a test about ordering, but the one about facetting by Date (probably it would not be a unit test).

**Comment by yutannihilation** [comment]
(But, since I don't have enough time to help right now and it seems not necessary for the release, please ignore my comments. Sorry...)

**Comment by paleolimbot** [comment]
I accidentally closed this from a typo in PR 3331...can I/should I fix this?

**Comment by yutannihilation** [comment]
BTW, what about `POSIXct`? This still fails with #3327.

``` r
library(ggplot2)

packageDescription("ggplot2")[c("GithubRef", "GithubSHA1")]
#> $GithubRef
#> [1] "plyr-order"
#> 
#> $GithubSHA1
#> [1] "b078f6901cc66968a04587585015559a42f71087"

df <- data.frame(date = as.Date("2019-01-01"), x = 1, y = 1)
df$date <- as.POSIXct(df$date)
p <- ggplot(df, aes(x, y)) + geom_point()
p + facet_grid(vars(date))
#> Error in scale_apply(layer_data, x_vars, "train", SCALE_X, x_scales):
```

![](https://i.imgur.com/4iGg0VR.png)

<sup>Created on 2019-05-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

**Comment by thomasp85** [comment]
I've fixed `POSIXct`... that was the last of the classes that received special treatment in plyr so hopefully this will be it for weird facet failures

**Comment by yutannihilation** [comment]
Oh, thanks for fixing!

**Comment by thomasp85** [comment]
closed by #3327

**Comment by cf814** [comment]
This issue appears to occur still with POSIXct variables, but not with Dates - see attached sample data and below code. 

[outlierSummary.zip](https://github.com/tidyverse/ggplot2/files/3479001/outlierSummary.zip)

outlierSummary$week2 <- as.Date(outlierSummary$week, format = "%Y-%m-%d")

obc=ggplot(data=outlierSummary,aes(x=type,y=outpc,fill=type))+geom_bar(stat = "identity")+
  ylim(c(0,3))+geom_text(aes(label=outliers,x=type,y=outpc),position=position_dodge(width=0.9), size=4,vjust=0)

#With posixct
obc <- obc + facet_grid(week ~ period)+labs(x="Provider", y="Weekly Rate of Outliers (%)",fill="Provider", caption="Numbers above bars: Outlier counts")+theme_bw()

#With Date
obc2 <- obc + facet_grid(week2 ~ period)+labs(x="Provider", y="Weekly Rate of Outliers (%)",fill="Provider", caption="Numbers above bars: Outlier counts")+theme_bw()

**Comment by paleolimbot** [comment]
Could you open a new issue with a more minimal example? It saves us quite a bit of time if we can copy and paste a small section of code to recreate the error.
